### PR TITLE
WIP: updated to v0.104.0

### PR DESCRIPTION
**Pull request by muelli**
note this this fails for me:

```
npm info it worked if it ends with ok
npm verb cli [ '/app/bin/node', '/app/bin/npm', 'run', 'submodule' ]
npm info using npm@6.4.1
npm info using node@v10.15.3
npm verb run-script [ 'presubmodule', 'submodule', 'postsubmodule' ]
npm info lifecycle deltachat-node@0.44.1~presubmodule: deltachat-node@0.44.1
npm info lifecycle deltachat-node@0.44.1~submodule: deltachat-node@0.44.1

> deltachat-node@0.44.1 submodule /run/build/delta/node_modules/deltachat-node
> git submodule update --recursive --init

fatal: not a git repository (or any parent up to mount point /run/build)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
npm verb lifecycle deltachat-node@0.44.1~submodule: unsafe-perm in lifecycle true
npm verb lifecycle deltachat-node@0.44.1~submodule: PATH: /app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/run/ccache/bin:/app/bin:/usr/bin
npm verb lifecycle deltachat-node@0.44.1~submodule: CWD: /run/build/delta/node_modules/deltachat-node
npm info lifecycle deltachat-node@0.44.1~submodule: Failed to exec submodule script
npm verb stack Error: deltachat-node@0.44.1 submodule: `git submodule update --recursive --init`
npm verb stack Exit status 128
npm verb stack     at EventEmitter.<anonymous> (/app/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:301:16)
npm verb stack     at EventEmitter.emit (events.js:189:13)
npm verb stack     at ChildProcess.<anonymous> (/app/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
npm verb stack     at ChildProcess.emit (events.js:189:13)
npm verb stack     at maybeClose (internal/child_process.js:970:16)
npm verb stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
npm verb pkgid deltachat-node@0.44.1
npm verb cwd /run/build/delta/node_modules/deltachat-node
npm verb Linux 4.9.0-8-amd64
npm verb argv "/app/bin/node" "/app/bin/npm" "run" "submodule"
npm verb node v10.15.3
npm verb npm  v6.4.1
npm ERR! code ELIFECYCLE
npm ERR! errno 128
npm ERR! deltachat-node@0.44.1 submodule: `git submodule update --recursive --init`
npm ERR! Exit status 128
npm ERR! 
npm ERR! Failed at the deltachat-node@0.44.1 submodule script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm verb exit [ 128, true ]
npm WARN Local package.json exists, but node_modules missing, did you mean to install?
npm timing npm Completed in 314ms

npm ERR! A complete log of this run can be found in:
npm ERR!     /run/build/delta/npm-cache/_logs/2019-06-09T17_48_43_361Z-debug.log
npm verb lifecycle deltachat-node@0.44.1~rebuild-core: unsafe-perm in lifecycle true
npm verb lifecycle deltachat-node@0.44.1~rebuild-core: PATH: /app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/app/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/run/build/delta/node_modules/deltachat-node/node_modules/.bin:/run/build/delta/node_modules/.bin:/run/ccache/bin:/app/bin:/usr/bin
```

It seems to be wanting to execute `git submodule update --recursive --init` which fails, of course.

I don't think we can do anything ourselves here. @Jikstra do you know how to not make it reach out to the Internet?

### CODE REVIEW

**Comment by Jikstra** [comment]
@muelli this happens if the install scripts tries to get the deltachat-core source code to compile. What's weird is that we didn't change a lot on that part since the 0.103.0 release. Not sure what's making trouble there.

**Comment by flub** [comment]
The updates to dc-node allow building against a system installed libdeltachat.so.  I think the .yml file needs to add a build step to install deltachat-core first so that there's a valid libdeltachat.so file and then add `--dc-system-lib=true` to the npm install command to stop it from downloading the core and use the installed core instead.

**Comment by flub** [comment]
Looking at `package.json` it seems the `npm run pack` command is also gone so that step will probably also fail.  I'm not familiar with what that has morphed into though, that'll need some digging.

**Comment by flub** [comment]
Ugh, ignore that last comment.  I mixed dc-node and dc-dekstop again.

**Comment by flub** [comment]
The `--dc-system-lib=true` thing is explained in the readme of dc-node (and you can see it implemented in package.json/install.js/bindings.gyp).  The flag should be passed along from dc-desktop to dc-node by npm though.

**Comment by muelli** [comment]
k.
thanks for the update.
I fail to build the rust module with 

```
   Compiling ryu v0.2.8
   Compiling subtle v2.1.0
error[E0554]: #![feature] may not be used on the stable release channel
  --> /run/build/deltachat-core-rust/cargo/vendor/subtle-2.1.0/src/lib.rs:12:34
   |
12 | #![cfg_attr(feature = "nightly", feature(asm))]
   |                                  ^^^^^^^^^^^^

error[E0554]: #![feature] may not be used on the stable release channel
  --> /run/build/deltachat-core-rust/cargo/vendor/subtle-2.1.0/src/lib.rs:13:34
   |
13 | #![cfg_attr(feature = "nightly", feature(external_doc))]
   |                                  ^^^^^^^^^^^^^^^^^^^^^

   Compiling openssl v0.10.23
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0554`.
error: Could not compile `subtle`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```
Or am I invoking the build wrongly?

File: chat.delta.desktop.yml [hunk 1]
```diff
@@ -70,6 +72,41 @@ modules:
         url: https://ftp.osuosl.org/pub/blfs/conglomeration/cyrus-sasl/cyrus-sasl-2.1.27.tar.gz
         sha256: 26866b1549b00ffd020f188a43c258017fa1c382b3ddadd8201536f72efb05d5
 
+  - name: deltachat-core-rust
+    buildsystem: simple
+    build-options:
+      env:
+        electron_config_cache: /run/build/delta/npm-cache
```
Comment (flub) [inline chat.delta.desktop.yml]: is this needed for rust?

**Comment by flub** [comment]
Sadly the rust code includes crypto code (rpgp) and crypto code implies assembler code to do stuff in constant time and other trickery, and in turn rust hasn't figured out how to expose asm code to be safe or at least how to provide a stable rust interface for it and thus they only support asm on nightly.

All this results in the fact we can't compile dcc-rs (rpgp really) on rust stable and must use rust nightly.  Last time I looked there was no nightly flatpak sdk like there is for stable.  I think it might be reasonable to replicate what the stable sdk does inside this yml file here, or maybe consider making a rust-nightly sdk?  But that seems like a lot of more work.

File: chat.delta.desktop.yml (hunk 1 above)
Comment (muelli) [reply chat.delta.desktop.yml]: nah. I just copy and pasted stuff. I intended to clean the commits up once it's working. Good catch, though!

**Comment by muelli** [comment]
hrm. Are you aware of any Rust plans to move the required features to the "stable" release?

There are on binary releases of rpgp or deltachat-core-rust that we could use instead, are there?

The description in the SDK https://github.com/flathub/org.freedesktop.Sdk.Extension.rust-stable/blob/master/org.freedesktop.Sdk.Extension.rust-stable.json#L25  looks relatively simple, but I'd want to avoid downloading the compiler :-/  An "unstable" Sdk sounds wrong to begin with and I guess people who built the stable SDK thought of providing an unstable one but discarded that idea.

**Comment by flub** [comment]
no, afaik rust won't move the required features to stable anytime soon because of the asm stuff i tried to explain a bit above.  The binary releases for rpgp don't work and anyway are only for the C ffi, I think for cargo you woundn't be able to use it so wouldn't help for dcc-rs (aka deltachat-core-rust).

I don't think anyone's worked on a binary release of dcc-rs either, though I'm somewhat surprised, would that meet the flatpak guidelines?

Anyway, short/medium term I think downloading the unstable rust compiler toolchain is the thing to do sadly.  Long term I'd hope rust figures out how to move asm to stable but I really don't know much about what's involved there and how likely that is.

**Comment by muelli** [comment]
Now building fails with this error: https://github.com/deltachat/deltachat-core-rust/issues/158

**Comment by flub** [comment]
I've created https://github.com/flathub/chat.delta.desktop/pull/21 which can be merged into this PR and gets a little further.

It's stuck on the generated npm sources being buggy, this probably warrants a bug report with the tool, i've tried to describe it a bit in the comment but didn't figure out a fix yet.

**Comment by flub** [comment]
btw, the latest releases where done with the old deltachat-core and rpgp afaik.  So while this is the way to do things in the future i'm not sure the flatpak should release with deltachat-core-rust just yet.  Anyway, this work is useful for the dcc+rpgp scenario too as things are largely similar.

**Comment by muelli** [comment]
I got a bit further by manually installing the deltachat.so and its header as well as pkg-config file.
It'd be good if there was a "make install" equivalent for deltachat-rust: https://github.com/deltachat/deltachat-core-rust/issues/170

Now it complains somewhere in deltachat-node:

```
gyp info spawn args [ 'V=1', 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory '/run/build/delta/node_modules/deltachat-node/build'
  cc '-DNODE_GYP_MODULE_NAME=deltachat' '-DUSING_UV_SHARED=1' '-DUSING_V8_SHARED=1' '-DV8_DEPRECATION_WARNINGS=1' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/app/include/node -I/app/src -I/app/deps/openssl/config -I/app/deps/openssl/openssl/include -I/app/deps/uv/include -I/app/deps/zlib -I/app/deps/v8/include -I../../napi-macros  -fPIC -pthread -Wall -Wextra -Wno-unused-parameter -m64 -std=gnu99 -I/app/include/deltachat -O3 -fno-omit-frame-pointer  -MMD -MF ./Release/.deps/Release/obj.target/deltachat/src/module.o.d.raw   -c -o Release/obj.target/deltachat/src/module.o ../src/module.c
../src/module.c: In function ‘dc_event_handler’:
../src/module.c:74:16: error: ‘DC_EVENT_HTTP_GET’ undeclared (first use in this function); did you mean ‘DC_EVENT_MSG_READ’?
   if (event == DC_EVENT_HTTP_GET) {
                ^~~~~~~~~~~~~~~~~
                DC_EVENT_MSG_READ
../src/module.c:74:16: note: each undeclared identifier is reported only once for each function it appears in
../src/module.c: In function ‘dcn_check_password’:
../src/module.c:412:16: warning: implicit declaration of function ‘dc_check_password’; did you mean ‘dcn_check_password’? [-Wimplicit-function-declaration]
   int result = dc_check_password(dcn_context->dc_context, password);
                ^~~~~~~~~~~~~~~~~
                dcn_check_password
make: *** [deltachat.target.mk:106: Release/obj.target/deltachat/src/module.o] Error 1
make: Leaving directory '/run/build/delta/node_modules/deltachat-node/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/app/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:189:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)

```

Is this a genuine bug in deltachat-node?

**Comment by flub** [comment]
Yep, that's a bug in dc-node, that HTTP_GET event no longer exists in dcc-rs.  Not sure about the so did it seems dc_check_password at least grepping the code doesn't show it anymore.

Thanks for all the work!

**Comment by flub** [comment]
https://github.com/deltachat/deltachat-core-rust/pull/172

This will leave installing libdeltachat.so and the deltachat.pc file to the packaging here, which is not the best end solution, but it at least makes it possible to use.  Making the rust lib install things into the right place is a larger conversation I guess.

**Comment by muelli** [comment]
reg. the build failure in the bot: https://github.com/flathub/flathub/issues/1058

**Comment by muelli** [comment]
> Yep, that's a bug in dc-node, that HTTP_GET event no longer exists in dcc-rs.

can you file one? I am a bit confused as to where to bug lies, because deltachat-node does not have that string. Could it be the version of deltachat-node that deltachat-desktop pulls in somehow?

**Comment by flub** [comment]
On Sat 06 Jul 2019 at 14:23 -0700, muelli wrote:

>> Yep, that's a bug in dc-node, that HTTP_GET event no longer exists in dcc-rs.
>
> can you file one? I am a bit confused as to where to bug lies, because
> deltachat-node does not have that string. Could it be the version of
> deltachat-node that deltachat-desktop pulls in somehow?

Hmm, maybe we got something wrong than.  I'll look into it.

**Comment by muelli** [comment]
I guess this is superseded by https://github.com/flathub/chat.delta.desktop/pull/23

**Pull request closed without merging**